src/inference/AdvancedSMC.jl

```diff
@@ -29,13 +29,123 @@ end
+#     selector::Selector
+#     vi::VarInfo
+# end
+
+mutable struct SMCState <: SamplerState
+    logevidence::Vector{Float64}
+    vi::VarInfo
+end
+
+SMCState() = SMCState(Float64[], VarInfo())
+
 function Sampler(alg::SMC, s::Selector)
-    info = Dict{Symbol, Any}()
-    info[:logevidence] = []
-    return Sampler(alg, info, s)
+    dict = Dict{Symbol, Any}()
+    state = SMCState()
+    return Sampler{SMC,SMCState}(alg, dict, s, state)
+end
+
+struct SMCTransition{T} <: AbstractTransition
+    names::Vector{Symbol}
+    θ::T
+end
+
+vnames(vi::VarInfo) = Symbol.(collect(keys(vi)))
+SMCTransition(vi::VarInfo) = SMCTransition{typeof(vi.vals)}(vnames(vi), vi.value)
+
+# VarInfo, combined with spl.info, to Sample
+function transition(vi::AbstractVarInfo, spl::Sampler{<:SMC})
+    # Array version
+    vals = vi.vals
+    nms = vnames(vi)
+    new_trans = SMCTransition{typeof(vals)}(nms, vals)
+    return new_trans
+
+
+    ### Namedtuple version.
+    # vals = vi.vals
+    # ks = collect(keys(vi))
+    # ln = length(ks)
+    # nms = ntuple(i -> Symbol(ks[i]), ln)
+    # vs = ntuple(i -> vals[i], ln)
+    # tup = NamedTuple{nms}(vs)
+    # println(tup)
+    # return s
+end
+
+#####################
+# Interface for SMC #
+#####################
+function sample_init!(
+    ::AbstractRNG,
+    ℓ::ModelType,
+    s::Sampler{SMC, SMCState},
+    N::Integer;
+    kwargs...
+) where {ModelType<:Sampleable}
+    # Do nothing.
+end
+
+function transitions_init(
+    ::AbstractRNG,
+    ℓ::ModelType,
+    s::Sampler{SMC, SMCState},
+    N::Integer;
+    kwargs...
+) where {ModelType<:Sampleable, SamplerType<:AbstractSampler}
+    return Vector{SMCTransition}(undef, N)
+end
+
+function step!(
+    a::AbstractRNG, # Note: This function does not use the range argument.
+    ℓ::Turing.Model,
+    s::Sampler{SMC, SMCState},
+    N::Integer, # Note: This function doesn't use the N argument.
+    ::T; # Note: No use of the previous draw.
+    kwargs...
+) where T<:AbstractTransition
+    return step!(a, ℓ, s, N; kwargs...)
+end
+
+function step!(
+    ::AbstractRNG, # Note: This function does not use the range argument.
+    ℓ::Turing.Model,
+    s::Sampler{SMC, SMCState},
+    ::Integer; # Note: This function doesn't use the N argument.
+    kwargs...
+)
+    particles = ParticleContainer{Trace}(ℓ)
+    s.state.vi.num_produce = 0;  # Reset num_produce before new sweep\.
+    set_retained_vns_del_by_spl!(s.state.vi, s)
+    resetlogp!(s.state.vi)
+
+    push!(particles, s.alg.n_particles, s, s.state.vi)
+
+    while consume(particles) != Val{:done}
+      ess = effectiveSampleSize(particles)
+      if ess <= s.alg.resampler_threshold * length(particles)
+        resample!(particles,s.alg.resampler)
+      end
+    end
+
+    ## pick a particle to be retained.
+    Ws, _ = weights(particles)
+    indx = randcat(Ws)
+    push!(s.state.logevidence, particles.logE)
+
+    params = particles[indx].vi
+
+    # update the master vi.
+    s.state.vi = params
```

Comment:
Can someone confirm that this step I've added is actually necessary? Basically, I update the main sampler state to include the `VarInfo` just received from a particle. Any problems with this? I think it could be excluded without issue, but I'm not sure.

Reply:
which variable do you store in `MCMCChains`, `s.state.vi` or `params`? If `params` is used, then this is not necessary, but won't cause a problem either since SMC samplers are stateless.

Reply:
Ultimately `params.vals` makes it to `MCMCChains`.